api/resources_portal/views/user.py

```diff
@@ -139,102 +167,81 @@ def get_permissions(self):
+                    "missing_parameters": missing_params,
+                    "malformed_grants": malformed_grants,
+                },
+                status=400,
             )
+
         if User.objects.filter(orcid=request.data["orcid"]).exists():
             return JsonResponse(
-                {"error": "A user with the provided ORCID already exists."}, status=400,
+                {
+                    "error": "A user with the provided ORCID already exists.",
+                    "error_code": "USER_EXISTS",
+                },
+                status=400,
             )
 
         try:
             api = orcid.PublicAPI(CLIENT_ID, CLIENT_SECRET, sandbox=IS_OAUTH_SANDBOX)
 
-            summary = api.read_record_public(
+            orcid_response = api.read_record_public(
                 request.data["orcid"], "record", request.data["access_token"]
             )
+            orcid_data = parse_orcid_response(orcid_response)
         except Exception as error:
             return JsonResponse({"error": error}, status=500,)
 
-        email = ""
-
-        if "email" in request.data:
-            email = request.data["email"]
-        else:
-            if len(summary["person"]["emails"]["email"]) == 0:
-                return JsonResponse(
-                    {
-                        "error": "There were no emails made availible on the provided ORCID record. Please provide an email in the POST request.",
-                        "needs_email": True,
-                    },
-                    status=401,
-                )
-
-            # Use the email first added to the ORCID account
-            email = summary["person"]["emails"]["email"][0]["email"]
+        # use request data fallback to parsed orcid data
+        user_data = {
+            "orcid": orcid_data["orcid"],
+            "orcid_access_token": request.data.get("access_token", None),
+            "orcid_refresh_token": request.data.get("refresh_token", None),
+            "first_name": request.data.get("first_name", orcid_data["first_name"]),
+            "last_name": request.data.get("last_name", orcid_data["last_name"]),
+            "email": request.data.get("email", orcid_data["email"]),
+        }
 
-        # Get first and last name
+        # check if missing param should be passed up
+        missing_data = get_unset_keys(user_data, ["email", "first_name", "last_name"])
 
-        first_name = summary["person"]["name"]["given-names"]["value"]
-        last_name = ""
-        if summary["person"]["name"].get("family-name", None):
-            last_name = summary["person"]["name"]["family-name"]["value"]
+        if missing_data:
+            return JsonResponse(
+                {
+                    "error": "There were details not provided on the provided ORCID record. Please provide missing details in the POST request.",
+                    "required": missing_data,
+                },
+                status=401,
```

Comment:
I think a 400 Bad Request makes more sense here than a 401 Unauthorized.

Reply:
updated to use 409 when user exists and 422 when orcid response is missing required info